hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AFailureHandling.java

```diff
@@ -72,6 +76,40 @@ public void testMultiObjectDeleteNoFile() throws Throwable {
     removeKeys(getFileSystem(), "ITestS3AFailureHandling/missingFile");
   }
 
+  /**
+   * See HADOOP-18112.
+   */
+  @Test
+  public void testMultiObjectDeleteLargeNumKeys() throws Exception {
+    S3AFileSystem fs =  getFileSystem();
+    Path path = path("largeDir");
+    mkdirs(path);
+    createFiles(fs, path, 1, 1005, 0);
+    List<String> keys = new ArrayList<>();
+    RemoteIterator<LocatedFileStatus> locatedFileStatusRemoteIterator =
+            fs.listFiles(path, false);
+    while (locatedFileStatusRemoteIterator.hasNext()) {
+      Path file = locatedFileStatusRemoteIterator.next().getPath();
+      keys.add(fs.pathToKey(file));
+    }
+    // After implementation of paging during multi object deletion,
+    // no exception is encountered.
+    Long bulkDeleteReqBefore = getNumberOfBulkDeleteRequestsMadeTillNow(fs);
```

Comment:
I was going to suggest adding this to `ITestS3ADeleteCost` , but you want to get at the private method, don't you?

Reply:
I don't understand this. Added here as other delete tests were present here only.

Reply:
I mean we do tests there which measure how many IO calls we made. but its in a different package to hard to pick up.

Reply:
Oh Okay. I think it is good to add this one here.